kv/kvnemesis: TestKVNemesisSingleNode_ReproposalChaos failed

kv/kvnemesis.TestKVNemesisSingleNode_ReproposalChaos [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/14690864?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/14690864?buildTab=artifacts#/) on master @ [972be0c2be8f85d9de791eecbd468f25be92f082](https://github.com/cockroachdb/cockroach/commits/972be0c2be8f85d9de791eecbd468f25be92f082):


```
=== RUN   TestKVNemesisSingleNode_ReproposalChaos
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/4aeb9c412c16a003ed8ddc43788d2bba/logTestKVNemesisSingleNode_ReproposalChaos1401109037
    test_log_scope.go:81: use -show-logs to present logs inline
    kvnemesis_test.go:328: seed: 5438976572559527631
    kvnemesis_test.go:235: kvnemesis logging to /artifacts/tmp/_tmp/4aeb9c412c16a003ed8ddc43788d2bba/kvnemesis982756613
    kvnemesis.go:178: error applying x.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
          txn.SetIsoLevel(isolation.Snapshot)
          txn.ReverseScanForUpdateSkipLocked(ctx, tk(4222404943025411556), tk(8969248248764761760), 0) // @1712188216.005474154,0 <nil>
          txn.CreateSavepoint(ctx, 1) // <nil>
          txn.ScanForShareSkipLocked(ctx, tk(2627400274596218044), tk(3567820568820274134), 0) // @1712188216.005474154,0 (/Table/100/"2930d9982c1ca220":v1, /Table/100/"2ec7005f2add7caf":v1, <nil>)
          txn.Get(ctx, tk(1233149570614889857)) // @1712188216.005474154,0 (<nil>, <nil>)
          txn.RollbackSavepoint(ctx, 1) // <nil>
          txn.ReverseScanForUpdateGuaranteedDurability(ctx, tk(7728008165367397093), tk(10576023797256947185), 0) // @1712188216.005474154,0 <nil>
          txn.ScanForUpdateSkipLocked(ctx, tk(1628354040708553529), tk(4838956009624024228), 0) // @1712188216.005474154,0 (/Table/100/"246498feb5763e34":v1, /Table/100/"2930d9982c1ca220":v1, /Table/100/"2ec7005f2add7caf":v1, <nil>)
          txn.GetForUpdateSkipLocked(ctx, tk(16960950265812106801)) // @1712188216.005474154,0 (<nil>, <nil>)
          txn.ReverseScanForShareSkipLockedGuaranteedDurability(ctx, tk(1401713875290176120), tk(7785689852733970154), 0) // @1712188216.005474154,0 (/Table/100/"33bd89017b7c914e":v1, /Table/100/"2ec7005f2add7caf":v1, /Table/100/"2930d9982c1ca220":v1, /Table/100/"246498feb5763e34":v1, <nil>)
          {
            b := &kv.Batch{}
            b.Get(tk(3617741942969719995)) // QueryIntent request for lock at sequence number 0 but sequence number is ignored [{0 0}]
            b.ReverseScanForShare(tk(4566508882259383921), tk(6114779437940260227)) // QueryIntent request for lock at sequence number 0 but sequence number is ignored [{0 0}]
            b.ReverseScanForShareGuaranteedDurability(tk(582022874963125873), tk(8507304394499687153)) // QueryIntent request for lock at sequence number 0 but sequence number is ignored [{0 0}]
            txn.Run(ctx, b) // QueryIntent request for lock at sequence number 0 but sequence number is ignored [{0 0}]
          }
          txn.GetForUpdate(ctx, tk(5338060224006893485)) // omitted
          txn.ScanForShare(ctx, tk(8497723395767383015), tk(16637431305443591149), 0) // omitted
          txn.GetForUpdateSkipLockedGuaranteedDurability(ctx, tk(9795111736106540646)) // omitted
          b := &kv.Batch{}
          b.ReverseScanForUpdateGuaranteedDurability(tk(12341411153894795778), tk(14630587199611829575)) // omitted
          txn.CommitInBatch(ctx, b) // omitted
          return nil
        }) // QueryIntent request for lock at sequence number 0 but sequence number is ignored [{0 0}]: QueryIntent request for lock at sequence number 0 but sequence number is ignored [{0 0}]
    kvnemesis.go:178: error applying x.Get(ctx, tk(3617741942969719995)) // QueryIntent request for lock at sequence number 0 but sequence number is ignored [{0 0}]: QueryIntent request for lock at sequence number 0 but sequence number is ignored [{0 0}]
    kvnemesis.go:178: error applying x.ReverseScanForShare(ctx, tk(4566508882259383921), tk(6114779437940260227), 0) // QueryIntent request for lock at sequence number 0 but sequence number is ignored [{0 0}]: QueryIntent request for lock at sequence number 0 but sequence number is ignored [{0 0}]
    kvnemesis.go:178: error applying x.ReverseScanForShareGuaranteedDurability(ctx, tk(582022874963125873), tk(8507304394499687153), 0) // QueryIntent request for lock at sequence number 0 but sequence number is ignored [{0 0}]: QueryIntent request for lock at sequence number 0 but sequence number is ignored [{0 0}]
    kvnemesis.go:213: failures(verbose): /artifacts/tmp/_tmp/4aeb9c412c16a003ed8ddc43788d2bba/kvnemesis982756613/failures
        repro steps: /artifacts/tmp/_tmp/4aeb9c412c16a003ed8ddc43788d2bba/kvnemesis982756613/repro.go
        rangefeed KVs: /artifacts/tmp/_tmp/4aeb9c412c16a003ed8ddc43788d2bba/kvnemesis982756613/kvs-rangefeed.txt
        scan KVs: /artifacts/tmp/_tmp/4aeb9c412c16a003ed8ddc43788d2bba/kvnemesis982756613/kvs-scan.txt
    kvnemesis_test.go:356: [0] proposed: 722
    kvnemesis_test.go:358: [0] reproposed unchanged: 58
    kvnemesis_test.go:360: [0] reproposed with new LAI: 66
    kvnemesis_test.go:365: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:365
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:290
        	Error:      	Should be zero, but was 4
        	Test:       	TestKVNemesisSingleNode_ReproposalChaos
        	Messages:   	kvnemesis detected failures
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/4aeb9c412c16a003ed8ddc43788d2bba/logTestKVNemesisSingleNode_ReproposalChaos1401109037
--- FAIL: TestKVNemesisSingleNode_ReproposalChaos (20.26s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #121426 kv/kvnemesis: TestKVNemesisSingleNode_ReproposalChaos failed [C-test-failure O-robot P-0 T-kv X-duplicate branch-release-24.1 release-blocker]
- #115598 kv/kvnemesis: TestKVNemesisSingleNode_ReproposalChaos failed [A-kv-transactions C-bug C-test-failure O-robot P-1 T-kv branch-release-23.2]
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestKVNemesisSingleNode_ReproposalChaos.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-37488

kv/kvnemesis.TestKVNemesisSingleNode_ReproposalChaos [failed](https://tanzanite.cluster.engflow.com/invocations/default/5b65eb98-ea00-4935-9667-b112196a8e83?testReportRun=10&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZuZW1lc2lzOmt2bmVtZXNpc190ZXN0) on master @ [0d3c5b71415de867fedde97de01486d424209848](https://github.com/cockroachdb/cockroach/commits/0d3c5b71415de867fedde97de01486d424209848):


```
=== RUN   TestKVNemesisSingleNode_ReproposalChaos
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestKVNemesisSingleNode_ReproposalChaos1035996976
    test_log_scope.go:81: use -show-logs to present logs inline
    kvnemesis_test.go:328: seed: 1846452384636116561
    kvnemesis_test.go:235: kvnemesis logging to /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvnemesis/kvnemesis_test/run_10_of_25/test.outputs/kvnemesis1165143236
    kvnemesis.go:178: error applying x.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
          txn.SetIsoLevel(isolation.Snapshot)
          txn.CreateSavepoint(ctx, 0) // <nil>
          txn.GetForShare(ctx, tk(10891681431522639908)) // @1712206310.579469942,0 (v4, <nil>)
          txn.ReverseScanForShare(ctx, tk(7350376747782103761), tk(7846325779404258075), 0) // @1712206310.579469942,0 <nil>
          txn.GetForUpdateSkipLockedGuaranteedDurability(ctx, tk(7881345654111761158)) // @1712206310.579469942,0 (<nil>, <nil>)
          txn.ScanForShareSkipLockedGuaranteedDurability(ctx, tk(858745554791835278), tk(11664614209722986057), 0) // @1712206310.579469942,0 (/Table/100/"8b0348ce1f6d6cc3":v2, /Table/100/"9727068e2ca00024":v4, <nil>)
          txn.ScanForUpdateSkipLockedGuaranteedDurability(ctx, tk(8926370398678533479), tk(10256293817551097702), 0) // @1712206310.579469942,0 <nil>
          txn.ReverseScan(ctx, tk(5141795980418869474), tk(13627611504489087007), 0) // @1712206310.579469942,0 (/Table/100/"b809463aaaddce65":v2, /Table/100/"9727068e2ca00024":v4, /Table/100/"8b0348ce1f6d6cc3":v2, <nil>)
          txn.Get(ctx, tk(2434183244087443096)) // @1712206310.579469942,0 (<nil>, <nil>)
          txn.Del(ctx, tk(2434183244087443096) /* @s6 */) // @1712206310.579469942,0 <nil>
          txn.GetForUpdateSkipLockedGuaranteedDurability(ctx, tk(2434183244087443096)) // @1712206310.579469942,0 (<nil>, <nil>)
          {
            b := &kv.Batch{}
            txn.Run(ctx, b) // <nil>
          }
          txn.GetForShareSkipLockedGuaranteedDurability(ctx, tk(2434183244087443096)) // @1712206310.579469942,0 (<nil>, <nil>)
          txn.GetForUpdateSkipLockedGuaranteedDurability(ctx, tk(4688817267042000398)) // @1712206310.579469942,0 (<nil>, <nil>)
          txn.RollbackSavepoint(ctx, 0) // <nil>
          txn.ScanForUpdateGuaranteedDurability(ctx, tk(9331722684421405963), tk(14751081052723379566), 0) // @1712206310.579469942,0 (/Table/100/"8b0348ce1f6d6cc3":v2, /Table/100/"9727068e2ca00024":v4, /Table/100/"b809463aaaddce65":v2, <nil>)
          txn.Get(ctx, tk(2434183244087443096)) // @1712206310.579469942,0 (<nil>, <nil>)
          txn.Get(ctx, tk(2746710087154959722)) // @1712206310.579469942,0 (<nil>, <nil>)
          txn.GetForShareSkipLockedGuaranteedDurability(ctx, tk(13724492467464107697)) // @1712206310.579469942,0 (<nil>, <nil>)
          return nil
        }) // QueryIntent request for lock at sequence number 2 but sequence number is ignored [{1 2}]: QueryIntent request for lock at sequence number 2 but sequence number is ignored [{1 2}]
    kvnemesis.go:213: failures(verbose): /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvnemesis/kvnemesis_test/run_10_of_25/test.outputs/kvnemesis1165143236/failures
        repro steps: /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvnemesis/kvnemesis_test/run_10_of_25/test.outputs/kvnemesis1165143236/repro.go
        rangefeed KVs: /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvnemesis/kvnemesis_test/run_10_of_25/test.outputs/kvnemesis1165143236/kvs-rangefeed.txt
        scan KVs: /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvnemesis/kvnemesis_test/run_10_of_25/test.outputs/kvnemesis1165143236/kvs-scan.txt
    kvnemesis_test.go:356: [0] proposed: 523
    kvnemesis_test.go:358: [0] reproposed unchanged: 63
    kvnemesis_test.go:360: [0] reproposed with new LAI: 27
    kvnemesis_test.go:365: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:365
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:290
        	Error:      	Should be zero, but was 1
        	Test:       	TestKVNemesisSingleNode_ReproposalChaos
        	Messages:   	kvnemesis detected failures
    panic.go:626: -- test log scope end --
--- FAIL: TestKVNemesisSingleNode_ReproposalChaos (15.11s)
test logs left over in: outputs.zip/logTestKVNemesisSingleNode_ReproposalChaos1035996976
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=10</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #121426 kv/kvnemesis: TestKVNemesisSingleNode_ReproposalChaos failed [C-test-failure O-robot P-0 T-kv X-duplicate branch-release-24.1 release-blocker]
- #115598 kv/kvnemesis: TestKVNemesisSingleNode_ReproposalChaos failed [A-kv-transactions C-bug C-test-failure O-robot P-1 T-kv branch-release-23.2]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestKVNemesisSingleNode_ReproposalChaos.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

--- a/pkg/kv/kvclient/kvcoord/txn_coord_sender_savepoints.go
+++ b/pkg/kv/kvclient/kvcoord/txn_coord_sender_savepoints.go
@@ -91,7 +91,9 @@ func (tc *TxnCoordSender) CreateSavepoint(ctx context.Context) (kv.SavepointToke
 	// TODO(nvanbenschoten): once #113765 is resolved, we should make this
 	// unconditional and push it into txnSeqNumAllocator.createSavepointLocked.
 	if tc.interceptorAlloc.txnPipeliner.hasAcquiredLocks() {
-		tc.interceptorAlloc.txnSeqNumAllocator.stepWriteSeqLocked()
+		if err := tc.interceptorAlloc.txnSeqNumAllocator.stepWriteSeqLocked(ctx); err != nil {
+			return nil, err
+		}
 	}
 
 	s := &savepoint{
@@ -159,7 +161,9 @@ func (tc *TxnCoordSender) RollbackToSavepoint(ctx context.Context, s kv.Savepoin
 			enginepb.IgnoredSeqNumRange{
 				Start: sp.seqNum, End: tc.interceptorAlloc.txnSeqNumAllocator.writeSeq,
 			})
-		tc.interceptorAlloc.txnSeqNumAllocator.stepWriteSeqLocked()
+		if err := tc.interceptorAlloc.txnSeqNumAllocator.stepWriteSeqLocked(ctx); err != nil {
+			return err
+		}
 	}
 
 	return nil

--- a/pkg/kv/kvclient/kvcoord/txn_interceptor_seq_num_allocator.go
+++ b/pkg/kv/kvclient/kvcoord/txn_interceptor_seq_num_allocator.go
@@ -83,6 +83,10 @@ type txnSeqNumAllocator struct {
 func (s *txnSeqNumAllocator) SendLocked(
 	ctx context.Context, ba *kvpb.BatchRequest,
 ) (*kvpb.BatchResponse, *kvpb.Error) {
+	if err := s.checkReadSeqNotIgnoredLocked(ba); err != nil {
+		return nil, kvpb.NewError(err)
+	}
+
 	for _, ru := range ba.Requests {
 		req := ru.GetInner()
 		oldHeader := req.Header()
@@ -99,8 +103,7 @@ func (s *txnSeqNumAllocator) SendLocked(
 		// Notably, this includes Get/Scan/ReverseScan requests that acquire
 		// replicated locks, even though they go through raft.
 		if kvpb.IsIntentWrite(req) || req.Method() == kvpb.EndTxn {
-			s.stepWriteSeqLocked()
-			if err := s.maybeAutoStepReadSeqLocked(ctx); err != nil {
+			if err := s.stepWriteSeqLocked(ctx); err != nil {
 				return nil, kvpb.NewError(err)
 			}
 			oldHeader.Sequence = s.writeSeq
@@ -172,8 +175,20 @@ func (s *txnSeqNumAllocator) stepReadSeqLocked(ctx context.Context) error {
 }
 
 // stepWriteSeqLocked increments the write seqnum.
-func (s *txnSeqNumAllocator) stepWriteSeqLocked() {
+func (s *txnSeqNumAllocator) stepWriteSeqLocked(ctx context.Context) error {
 	s.writeSeq++
+	return s.maybeAutoStepReadSeqLocked(ctx)
+}
+
+// checkReadSeqNotIgnoredLocked verifies that the read seqnum is not in the
+// ignored seqnum list of the provided batch request.
+func (s *txnSeqNumAllocator) checkReadSeqNotIgnoredLocked(ba *kvpb.BatchRequest) error {
+	if enginepb.TxnSeqIsIgnored(s.readSeq, ba.Txn.IgnoredSeqNums) {
+		return errors.AssertionFailedf(
+			"read sequence number %d but sequence number is ignored %v after savepoint rollback",
+			s.readSeq, ba.Txn.IgnoredSeqNums)
+	}
+	return nil
 }
 
 // configureSteppingLocked configures the stepping mode.

--- a/pkg/kv/kvclient/kvcoord/txn_interceptor_seq_num_allocator_test.go
+++ b/pkg/kv/kvclient/kvcoord/txn_interceptor_seq_num_allocator_test.go
@@ -206,7 +206,7 @@ func TestSequenceNumberAllocationWithStep(t *testing.T) {
 	currentStepSeqNum := s.writeSeq
 
 	ba := &kvpb.BatchRequest{}
-	ba.Requests = nil
+	ba.Header = kvpb.Header{Txn: &txn}
 	ba.Add(&kvpb.ConditionalPutRequest{RequestHeader: kvpb.RequestHeader{Key: keyA}})
 	ba.Add(&kvpb.GetRequest{RequestHeader: kvpb.RequestHeader{Key: keyA}})
 	ba.Add(&kvpb.InitPutRequest{RequestHeader: kvpb.RequestHeader{Key: keyA}})

